Add Time-Series Gapfilling functionality.

weixiangsun opened the pull request:
## Description
The data set can be time series data. We might need to aggregate the data per time bucket. The data can be missing per entity inside the time bucket. So we need gap fill the missing data per entity for the time buckets before doing the aggregation.

Design: #7422

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
<!-- If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release. -->

<!-- If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text.
-->
## Documentation
<!-- If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document
-->

mayankshriv commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -117,21 +117,50 @@ private static String removeTerminatingSemicolon(String sql) {
-    PinotQuery pinotQuery = compileCalciteSqlToPinotQuery(sql);
+    PinotQuery pinotQuery = compileSqlNodeToPinotQuery(sqlNode);
+
+    if (enablePreAggregateGapfillQuery) {
```
Can we avoid this special-casing? We already have the `IN_SUBQUERY` feature: https://docs.pinot.apache.org/users/user-guide-query/filtering-with-idset#in_subquery

weixiangsun replied:
We can not leverage this IN_SUBQUERY feature https://docs.pinot.apache.org/users/user-guide-query/filtering-with-idset#in_subquery since the query syntax is trying to follow the real subquery approach.

The reason of this special-casing is that we are not blocking subquery feature development. We are planning to make our feature compatible with subquery feature when it is in place.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/plan/PreAggGapFillSelectionPlanNode.java:
```diff
@@ -0,0 +1,86 @@
+
+    List<ExpressionContext> args = gapFillSelection.getFunction().getArguments();
+    ExpressionContext timeSeriesOn = null;
+    for (int i = 5; i < args.size(); i++) {
```
The use of `5` hear reads like a magic number. Can we make it more readable and/or add comments?

weixiangsun replied:
Done

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/BlockValSetImpl.java:
```diff
@@ -0,0 +1,171 @@
+ * Helper class to convert the result rows to BlockValSet.
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class BlockValSetImpl implements BlockValSet {
```
Could we choose a better name for the class and also add more java docs. For example, unclear to me how it is different from other impls of `BlockValSet` without having to read the code.

weixiangsun replied:
Done

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/BlockValSetImpl.java:
```diff
@@ -0,0 +1,171 @@
+      }
+      return result;
+    }
+    throw new UnsupportedOperationException("Not supported");
```
IIRC, other parts of the code allow for read `int` as `long` and other such upcasting?

weixiangsun replied:
Should we allow to cast Integer to Long or not?

Jackie-Jiang commented on pinot-common/src/thrift/query.thrift:
```diff
@@ -20,6 +20,7 @@ namespace java org.apache.pinot.common.request
 
 struct DataSource {
   1: optional string tableName;
+  2: optional PinotQuery preAggregateGapfillQuery;
```
Let's name it `subquery`. We want it to be general in the query syntax

weixiangsun replied:
Done

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -117,21 +117,50 @@ private static String removeTerminatingSemicolon(String sql) {
   public static PinotQuery compileToPinotQuery(String sql)
       throws SqlCompilationException {
-    // Remove the comments from the query
-    sql = removeComments(sql);
```
This should not be removed

weixiangsun replied:
This is from conflict resolution. Fixed.

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -117,21 +117,50 @@ private static String removeTerminatingSemicolon(String sql) {
       throws SqlCompilationException {
-    // Remove the comments from the query
-    sql = removeComments(sql);
+    return compileToPinotQuery(sql, false);
```
Let's directly extend this method to support subquery within the `FROM` clause. The subquery parsing should always be enabled regardless of the query type

weixiangsun replied:
Done

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:
```diff
@@ -85,6 +85,7 @@
   // Keep the BrokerRequest to make incremental changes
   // TODO: Remove it once the whole query engine is using the QueryContext
   private final BrokerRequest _brokerRequest;
+  private QueryContext _preAggregateGapFillQueryContext;
```
Can we generify this to `subquery`? I think this can be modeled as general subquery

siddharthteotia replied:
@Jackie-Jiang 

So this was one of the things which was discussed a lot. My concern was that changing the PinotQuery now to accommodate "**generic**" subquery has to be done carefully accounting for standard sql subquery syntax and semantics and we should be confident that it will hold in future

If we are really touching the FROM clause, my suggestion would be to make sure we understand calcite's treatment of simple FROM clause (table name as today) and complex FROM clause (sub-queries). Whatever we do today in the FROM clause to make this particular gapfill sub-query work should not interfere in the future when we are going to leverage / extend calcite's generic subquery planner to support all kinds of subqueries. 

So this is why the path of least resistance could be to not touch PinotQuery for generic Subquery support since that may require a lot of design thinking before agreeing upon how to model any subquery in Pinot. So may be making it specific like in the above code just for gapfill is the way to go ?

weixiangsun replied:
@siddharthteotia @Jackie-Jiang 
Q: If we are really touching the FROM clause, my suggestion would be to make sure we understand calcite's treatment of simple FROM clause (table name as today) and complex FROM clause (sub-queries).
A: From calcite's perspective, the simple FROM clause (table name only) will be compiled as SqlIdentifier. The complex FROM clause will be compiled as SqlOrderBy object or SqlSelect object depending on if there is any order by.

It might hold true in the future to make subquery as part of DataSource instead of making the particular gapfill sub-query.

I lean towards generic approach since 
1. it is consistent with the calcite
2. even if it might impact the subquery feature, it can be fixed as part of the subquery feature development. 

@siddharthteotia any other suggestion?

Jackie-Jiang replied:
@siddharthteotia From query syntax perspective, I don't see difference between pre-aggregate-gap-fill and general subquery. Essentially the `FROM` clause contains a query instead of a table name. I'd suggest adding it as general subquery, but we can support it in the future.

siddharthteotia replied:
> even if it might impact the subquery feature, it can be fixed as part of the subquery feature development.

This ^^ is what I am concerned about. We are making changes to the wire object. We will have to live with it and if the decision taken now somehow does not hold for generic subquery then we will find ourselves retro-fitting subquery onto what we do now.

Note that generic subquery does not only include subquery in FROM clause but also in WHERE and other parts of a subquery. I haven't spend full time figuring out how to model all kinds of subquery in Pinot so just being extra careful here.

I think generic subquery will work and will be more cleaner. So may be let's go with it

weixiangsun replied:
Done

Jackie-Jiang reviewed (commented):
Let's try to not add the pre-aggregate specific attributes into the general query syntax/context. It can be modeled as standard subquery

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/PreAggregationGapFillSelectionOperatorService.java:
```diff
@@ -0,0 +1,388 @@
+ * <p>Expected behavior:
+ * <ul>
+ *   <li>
+ *     Return selection results with the same order of columns as user passed in.
```
Make the comment match with the class

weixiangsun replied:
Done

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/PreAggregationGapFillSelectionOperatorService.java:
```diff
@@ -0,0 +1,388 @@
+    return epoch / sz * sz;
+  }
+
+  List<Object[]> gapFillAndAggregate(List<Object[]> sortedRows, DataSchema dataSchemaForAggregatedResult) {
```
break it into the small functions.

amrishlal commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -130,8 +130,25 @@ public static PinotQuery compileToPinotQuery(String sql)
       sql = removeOptionsFromSql(sql);
     }
+
+    SqlParser sqlParser = SqlParser.create(sql, PARSER_CONFIG);
+    SqlNode sqlNode;
+    try {
+      sqlNode = sqlParser.parseQuery();
+    } catch (SqlParseException e) {
+      throw new SqlCompilationException("Caught exception while parsing query: " + sql, e);
+    }
+
     // Compile Sql without OPTION statements.
-    PinotQuery pinotQuery = compileCalciteSqlToPinotQuery(sql);
+    PinotQuery pinotQuery = compileSqlNodeToPinotQuery(sqlNode);
+
+    SqlSelect sqlSelect = getSelectNode(sqlNode);
+    if (sqlSelect != null) {
+      SqlNode fromNode = sqlSelect.getFrom();
+      if (fromNode != null && (fromNode instanceof SqlSelect || fromNode instanceof SqlOrderBy)) {
+        pinotQuery.getDataSource().setSubquery(compileSqlNodeToPinotQuery(fromNode));
+      }
+    }
```
Is all this change to original code necessary? For example the original function `compileCalciteSqlToPinotQuery` could still be retained, but modified to set the subquery node in pinotQuery.

weixiangsun replied:
I just broke down compileCalciteSqlToPinotQuery into two parts:
1. converting sql statement to SqlNode
2. construct PinotQuery from SqlNode.
This breakdown is because we need reuse the second part for constructing the pinotQuery from outer query and inner query.
So the change to original code is necessary.

amrishlal commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -321,32 +338,30 @@ private static void setOptions(PinotQuery pinotQuery, List<String> optionsStatem
-    }
-
-    PinotQuery pinotQuery = new PinotQuery();
-    if (sqlNode instanceof SqlExplain) {
-      // Extract sql node for the query
-      sqlNode = ((SqlExplain) sqlNode).getExplicandum();
-      pinotQuery.setExplain(true);
-    }
```
We need this for EXPLAIN queries to work.

weixiangsun replied:
This piece of the code has been moved to somewhere else and it will be executed.

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/util/GapfillUtils.java:
```diff
@@ -31,7 +34,10 @@
  */
 public class GapfillUtils {
   private static final String POST_AGGREGATE_GAP_FILL = "postaggregategapfill";
+  private static final String PRE_AGGREGATE_GAP_FILL = "preaggregategapfill";
```
I am wondering if this function should be renamed to such gapfill since the "pre aggregate" part is now clearly visible with the subquery structure?

weixiangsun replied:
Done

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/plan/CombinePlanNode.java:
```diff
@@ -180,6 +181,8 @@ public BaseCombineOperator run() {
         // Selection order-by
         return new SelectionOrderByCombineOperator(operators, _queryContext, _executorService);
       }
+    } else if (GapfillUtils.isPreAggregateGapfill(_queryContext)) {
+        return new SelectionOnlyCombineOperator(operators, _queryContext, _executorService);
```
Will this still work if an ORDER BY clause is present in the gap fill query? If not then perhaps some validation checks should be done to filter out unsupported usage of gapfill queries (ordery by, having, group by etc.)

siddharthteotia commented:
@weixiangsun  is this no longer being worked on ?

weixiangsun commented:
@siddharthteotia I am actively working on it.

weixiangsun commented on pinot-core/src/test/java/org/apache/pinot/queries/PreAggregationGapfillQueriesTest.java:
```diff
@@ -0,0 +1,3088 @@
+        + "  FROM parkingData "
+        + "  WHERE eventTime >= 1636257600000 AND eventTime <= 1636286400000 "
+        + "  LIMIT 200 "
+        + ") "
```
We should be able to support alias for the subquery.

weixiangsun replied:
Create a work item and put the TODO. Will address it in separate PR.

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/operator/combine/GapfillGroupByOrderByCombineOperator.java:
```diff
@@ -0,0 +1,263 @@
+  public static final int MAX_TRIM_THRESHOLD = 1_000_000_000;
+  private static final Logger LOGGER = LoggerFactory.getLogger(GapfillGroupByOrderByCombineOperator.class);
+  private static final String OPERATOR_NAME = "GapfillGroupByOrderByCombineOperator";
+  private static final String EXPLAIN_NAME = "GAPFILL_COMBINE_GROUPBY_ORDERBY";
```
This name should be changed to COMBINE_GROUPBY_ORDERBY_GAPFILL to keep the naming scheme consistent with other Combine operators.

weixiangsun replied:
Done

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/operator/combine/GapfillGroupByOrderByCombineOperator.java:
```diff
@@ -0,0 +1,263 @@
+ *   all threads
+ */
+@SuppressWarnings("rawtypes")
+public class GapfillGroupByOrderByCombineOperator extends BaseCombineOperator {
```
Except for setting some variables in the constructor, this file appears to be a close copy of `GroupByOrderByCombineOperator`. I am wondering if it would be possible to extend `GapfillGroupByOrderByCombineOperator` from GroupByOrderByCombineOperator to avoid code duplication?

weixiangsun replied:
Removed

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/plan/GapfillAggregationGroupByOrderByPlanNode.java:
```diff
@@ -0,0 +1,110 @@
+ * group-by order-by query on a single segment.
+ */
+@SuppressWarnings("rawtypes")
+public class GapfillAggregationGroupByOrderByPlanNode implements PlanNode {
```
Would it be possible to extend this class from `AggregationGroupByOrderByPlanNode` for code reuse or do some refactoring which would allow both classes to share common code?

weixiangsun replied:
This class is removed due to query syntax change. Done

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/plan/GapfillAggregationPlanNode.java:
```diff
@@ -0,0 +1,175 @@
+ * a single segment.
+ */
+@SuppressWarnings("rawtypes")
+public class GapfillAggregationPlanNode implements PlanNode {
```
Would it be possible to extend this class from `AggregationPlanNode` or do some sort of refactoring that would allow both classes to share common code?

weixiangsun replied:
Removed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/QueryContextUtils.java:
```diff
@@ -38,7 +39,11 @@ private QueryContextUtils() {
+      return isSelectionOnlyQuery(query.getSubQueryContext());
+    } else {
+      return query.getAggregationFunctions() == null;
+    }
```
We can not assume that subquery exists then this is automatically a gapfill query since this will block subquery feature.

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/QueryContextUtils.java:
```diff
@@ -47,16 +52,22 @@ public static boolean isSelectionQuery(QueryContext query) {
-    return query.getAggregationFunctions() == null && query.getOrderByExpressions() == null;
+    return query.getAggregationFunctions() == null
+        && query.getOrderByExpressions() == null
+        && !GapfillUtils.isGapfill(query);
```
Done

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/QueryContextUtils.java:
```diff
@@ -47,16 +52,22 @@ public static boolean isSelectionQuery(QueryContext query) {
-    AggregationFunction[] aggregationFunctions = query.getAggregationFunctions();
-    return aggregationFunctions != null && (aggregationFunctions.length != 1
-        || !(aggregationFunctions[0] instanceof DistinctAggregationFunction));
+    if (GapfillUtils.isGapfill(query)) {
```
Done

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/QueryContextUtils.java:
```diff
@@ -38,7 +39,11 @@ private QueryContextUtils() {
    */
   public static boolean isSelectionQuery(QueryContext query) {
-    return query.getAggregationFunctions() == null;
+    if (GapfillUtils.isGapfill(query)) {
```
Done

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/util/GapfillUtils.java:
```diff
@@ -31,7 +36,25 @@
+  private static final int STARTING_INDEX_OF_OPTIONAL_ARGS_FOR_PRE_AGGREGATE_GAP_FILL = 5;
+
+  public enum GapfillType {
+    // one sql query with gapfill only
+    Gapfill,
+    // gapfill as subquery, the outer query may have the filter
+    GapfillSelect,
+    // gapfill as subquery, the outer query has the aggregation
+    GapfillAggregate,
+    // aggregation as subqery, the outer query is gapfill
+    AggregateGapfill,
+    // aggegration as second nesting subquery, gapfill as fist nesting subquery, different aggregation as outer query
+    AggregateGapfillAggregate,
+    // no gapfill at all.
+    None
```
DONE

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/plan/CombinePlanNode.java:
```diff
@@ -161,8 +162,18 @@ public BaseCombineOperator run() {
       // Streaming query (only support selection only)
       return new StreamingSelectionOnlyCombineOperator(operators, _queryContext, _executorService, _streamObserver);
     }
+    GapfillUtils.GapfillType gapfillType = GapfillUtils.getGapfillType(_queryContext);
```
The reason I need handle the gapfill query first because otherwise I need make sure the query is not gapfill query for last two ifs. I prefer to keep them as they are.

Jackie-Jiang reviewed (commented):
High level question: Since the gap-fill is always processed on the broker side, will it be simpler if we always rewrite the query on the broker side to trim off the gap-fill part, so that no change is required on the server side? Another benefit of this approach is that this can eliminate the overhead of parsing sub-queries and processing gap-fill again on the servers.

weixiangsun commented on pinot-core/src/test/java/org/apache/pinot/queries/PreAggregationGapfillQueriesTest.java:
```diff
@@ -0,0 +1,3277 @@
+        + "    '27270960',  '27271440', '1:HOURS',"
+        + "     FILL(occupied, 'FILL_PREVIOUS_VALUE'), TIMESERIESON(levelId, lotId)) AS time_col,"
+        + "     occupied, lotId, levelId"
+        + "  FROM ("
```
Yes we do not stop taking arbitrary level of queries.

weixiangsun commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -129,8 +129,25 @@ public static PinotQuery compileToPinotQuery(String sql)
-    PinotQuery pinotQuery = compileCalciteSqlToPinotQuery(sql);
+    PinotQuery pinotQuery = compileSqlNodeToPinotQuery(sqlNode);
+
+    SqlSelect sqlSelect = getSelectNode(sqlNode);
+    if (sqlSelect != null) {
+      SqlNode fromNode = sqlSelect.getFrom();
+      if (fromNode != null && (fromNode instanceof SqlSelect || fromNode instanceof SqlOrderBy)) {
+        pinotQuery.getDataSource().setSubquery(compileSqlNodeToPinotQuery(fromNode));
+      }
+    }
```
Good catch! Fixed

weixiangsun commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -325,32 +342,30 @@ private static void setOptions(PinotQuery pinotQuery, List<String> optionsStatem
       selectNode.setFetch(orderByNode.fetch);
       selectNode.setOffset(orderByNode.offset);
-    } else {
+    } else if (sqlNode instanceof SqlSelect) {
```
Revert the change. Fixed.

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/data/table/IndexedTable.java:
```diff
@@ -65,10 +66,15 @@ protected IndexedTable(DataSchema dataSchema, QueryContext queryContext, int res
 
-    List<ExpressionContext> groupByExpressions = queryContext.getGroupByExpressions();
+    List<ExpressionContext> groupByExpressions;
+    if (queryContext.getGapfillType() != GapfillUtils.GapfillType.NONE) {
```
Done

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/data/table/IndexedTable.java:
```diff
@@ -65,10 +66,15 @@ protected IndexedTable(DataSchema dataSchema, QueryContext queryContext, int res
+    } else {
+      groupByExpressions = queryContext.getGroupByExpressions();
+    }
+    _aggregationFunctions = queryContext.getAggregationFunctions();
```
Fixed.

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:
```diff
@@ -85,6 +86,7 @@
   // Keep the BrokerRequest to make incremental changes
   // TODO: Remove it once the whole query engine is using the QueryContext
   private final BrokerRequest _brokerRequest;
+  private QueryContext _subqueryContext;
```
_gapfillType is calculated when the whole queryContext is constructed. _subqueryContext is final now

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/PreAggregateGapfillFilterHandler.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * Handler for Filter clause of PreAggregateGapFill.
+ */
+public class PreAggregateGapfillFilterHandler implements ValueExtractorFactory {
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/filter/RowMatcher.java:
```diff
@@ -0,0 +1,49 @@
+  /**
+   * Helper method to construct a RowMatcher based on the given filter.
+   */
+  public static RowMatcher getRowMatcher(FilterContext filter, ValueExtractorFactory valueExtractorFactory) {
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/PreAggregationGapFillDataTableReducer.java:
```diff
@@ -0,0 +1,775 @@
+    Preconditions.checkArgument(findTimeBucket, "No Group By timebucket.");
+  }
+
+  private List<Object[]> aggregateGapfilledData(List<Object[]> bucketedRows, DataSchema dataSchema) {
```
IndexedTable is used to merge the Intermediate Aggregate Result. We do not need merge the intermediate aggregated result inside outer aggregation since there are no multiple segments. 
If the subquery is aggregation query, the intermediate result from different pinot segments has already been merged before gapfill happens. IndexedTable is used to merge the intermediate result there.

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/PreAggregateGapfillFilterHandler.java:
```diff
@@ -0,0 +1,74 @@
+    if (expression.getType() == ExpressionContext.Type.IDENTIFIER) {
+      return new ColumnValueExtractor(_indexes.get(expression.getIdentifier()), _dataSchema);
+    } else {
+      return new ColumnValueExtractor(_indexes.get(expression.getFunction().toString()), _dataSchema);
```
The dataschema here is generated by PreAggregationGapFillDataTableReducer. FunctionContext.toString() is used in this case.

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/plan/GapfillSelectionPlanNode.java:
```diff
@@ -0,0 +1,90 @@
+
+    QueryContext queryContext = getSelectQueryContext();
+    Preconditions.checkArgument(queryContext.getOrderByExpressions() == null,
+        "The gapfill query should not have orderby expression.");
```
At the beginning, I thought since the result will be sorted by timestamp anyway, maybe we should not allow the customer to order the result by any other dimension. But you might be right, we should allow the customer order the result by the other dimension(s) also. We do not need do the ordering on pinot server for now. We can do the sorting on the broker side. I will make the change.

weixiangsun replied:
Fixed

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/PreAggregationGapFillDataTableReducer.java:
```diff
@@ -0,0 +1,775 @@
+    } else {
+      try {
+        IndexedTable indexedTable = getIndexedTable(dataSchema, dataTableMap.values(), reducerContext);
+        sortedRawRows = mergeAndSort(indexedTable, dataSchema);
```
It seems like mergeAndSort can be made much much faster if the the servers pre-sorted their resultset before sending it to the broker. Servers would sort their resultset in parallel and in this case, one would only need to do a k-way merge here to produce a globally sorted resultset without actually doing a full global sort or all the records.

weixiangsun replied:
Fixed

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/PreAggregationGapFillDataTableReducer.java:
```diff
@@ -0,0 +1,775 @@
+    List<Object[]> resultRows;
+    replaceColumnNameWithAlias(dataSchema);
+    if (_queryContext.getAggregationFunctions() != null) {
+      validateGroupByForOuterQuery();
```
Why do we need to do this validation? If the _queryContext object was constructed properly, there should be no need to do this validation? Can this validation be moved to compilation stage?

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/plan/CombinePlanNode.java:
```diff
@@ -180,6 +191,8 @@ public BaseCombineOperator run() {
         return new SelectionOrderByCombineOperator(operators, _queryContext, _executorService);
       }
+    } else if (gapfillType != GapfillUtils.GapfillType.NONE) {
+        return new SelectionOnlyCombineOperator(operators, _queryContext, _executorService);
```
We do not need do the sorting on the pinot server. The sorting will be done broker side.

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/PreAggregationGapFillDataTableReducer.java:
```diff
@@ -0,0 +1,775 @@
+import org.apache.pinot.spi.data.DateTimeFormatSpec;
+import org.apache.pinot.spi.data.DateTimeGranularitySpec;
+
+/**
+ * Helper class to reduce and set Aggregation results into the BrokerResponseNative
+ */
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/PreAggregationGapFillDataTableReducer.java:
```diff
@@ -0,0 +1,775 @@
+ * Helper class to reduce and set Aggregation results into the BrokerResponseNative
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class PreAggregationGapFillDataTableReducer implements DataTableReducer {
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/PreAggregationGapFillDataTableReducer.java:
```diff
@@ -0,0 +1,775 @@
+
+    ExpressionContext gapFillSelection = GapfillUtils.getGapfillExpressionContext(queryContext);
+
+    Preconditions.checkArgument(
+        gapFillSelection != null && gapFillSelection.getFunction() != null, "Gapfill Expression should be function.");
+    List<ExpressionContext> args = gapFillSelection.getFunction().getArguments();
+    Preconditions.checkArgument(
+        args.size() > 5, "PreAggregateGapFill does not have correct number of arguments.");
+    Preconditions.checkArgument(
+        args.get(1).getLiteral() != null, "The second argument of PostAggregateGapFill should be TimeFormatter.");
+    Preconditions.checkArgument(
+        args.get(2).getLiteral() != null, "The third argument of PostAggregateGapFill should be start time.");
+    Preconditions.checkArgument(
+        args.get(3).getLiteral() != null, "The fourth argument of PostAggregateGapFill should be end time.");
+    Preconditions.checkArgument(
+        args.get(4).getLiteral() != null, "The fifth argument of PostAggregateGapFill should be time bucket size.");
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/plan/CombinePlanNode.java:
```diff
@@ -161,8 +162,18 @@ public BaseCombineOperator run() {
-      if (_queryContext.getGroupByExpressions() == null) {
+      if (gapfillType == GapfillUtils.GapfillType.AGGREGATE_GAP_FILL_AGGREGATE) {
+        _queryContext.getSubQueryContext().getSubQueryContext().setEndTimeMs(_queryContext.getEndTimeMs());
+        return new GroupByOrderByCombineOperator(
```
Fixed

weixiangsun commented:
> High level question: Since the gap-fill is always processed on the broker side, will it be simpler if we always rewrite the query on the broker side to trim off the gap-fill part, so that no change is required on the server side? Another benefit of this approach is that this can eliminate the overhead of parsing sub-queries and processing gap-fill again on the servers.

If the gap-fill part is trimmed off, it is hard to tell if this is gapfill query in case of Gapfill Type query. We do not parse the sub-queries and processing gap-fill again on servers.

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:
```diff
@@ -133,6 +137,8 @@ private QueryContext(String tableName, List<ExpressionContext> selectExpressions
     _queryOptions = queryOptions;
     _debugOptions = debugOptions;
     _brokerRequest = brokerRequest;
+    _gapfillType = null;
```
Instead of setting `_gapfillType` to null, can we set it to `GAP_FILL_NONE` as that will clearly show that no gap filling will happen by default?

weixiangsun replied:
GAP_FILL_NONE was my first version of implementation. @Jackie-Jiang proposed "null" option. I am fine with both solutions, but I do not want to make another change. I prefer to keep as it is.

@amrishlal Please let me know if you have strong opinion about it.

amrishlal replied:
Let's leave it as null as Jackie suggested.

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/BrokerRequestToQueryContextConverter.java:
```diff
@@ -53,12 +55,81 @@ private BrokerRequestToQueryContextConverter() {
+    if (brokerRequest.getPinotQuery() != null) {
+      QueryContext queryContext = convertSQL(brokerRequest.getPinotQuery(), brokerRequest);
+      queryContext.setGapfillType(GapfillUtils.getGapfillType(queryContext));
+      validateForGapfillQuery(queryContext);
```
It seems like this validation is just checking syntax, can it be done on the Broker? If I am not mistaken almost all of Pinot's syntax checks happen on the Broker side.

weixiangsun replied:
If I am not mistaken, BrokerRequestToQueryContextConverter is running inside Pinot Broker. @amrishlal can you clarify your comment further?

amrishlal replied:
From what I can see these syntax checks are happening on the Server side. Server receives an `InstanceRequest` from Broker in `InstanceRequestHandler.channelRead0` method and invokes the `ServerQueryRequest` constructor to create `ServerQueryRequest` object. The `ServerQueryRequest` constructor then calls makes the call:

    `_queryContext = BrokerRequestToQueryContextConverter.convert(brokerRequest);`

We should move the syntax checks as far up the stack as possible and reject bad queries as early as possible so that unnecessary processing can be avoided.

weixiangsun replied:
Fixed

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/util/GapfillUtils.java:
```diff
@@ -119,4 +130,138 @@ static public Serializable getDefaultValue(DataSchema.ColumnDataType dataType) {
+    if (queryContext.getSubQueryContext() == null) {
+      if (isGapfill(queryContext)) {
+        Preconditions.checkArgument(queryContext.getAggregationFunctions() == null,
+            "Aggregation and Gapfill can not be in the same sql statement.");
```
Would it be possible to do these syntax checks all together in one place on the broker side when the initial query is received? That way bad queries can quickly be filtered out and we don't have to do syntax checks deep inside code logic.

weixiangsun replied:
Fixed.

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/BrokerRequestToQueryContextConverter.java:
```diff
@@ -53,12 +55,81 @@ private BrokerRequestToQueryContextConverter() {
-    return brokerRequest.getPinotQuery() != null ? convertSQL(brokerRequest) : convertPQL(brokerRequest);
+    if (brokerRequest.getPinotQuery() != null) {
+      QueryContext queryContext = convertSQL(brokerRequest.getPinotQuery(), brokerRequest);
+      queryContext.setGapfillType(GapfillUtils.getGapfillType(queryContext));
```
Seems like this should be done using the `QueryContext.Builder` pattern inside of `convertSQL` function?

weixiangsun replied:
The GapfillType is decided when all nested query context(s) are constructed. It is hard to be decided before each QueryContext is constructed.

amrishlal commented on pinot-core/src/test/java/org/apache/pinot/queries/GapfillQueriesTest.java:
```diff
@@ -0,0 +1,3615 @@
+// TODO: Item 2. Deprecate PostAggregateGapfill implementation in next PR
+@SuppressWarnings("rawtypes")
+public class GapfillQueriesTest extends BaseQueriesTest {
+  private static final File INDEX_DIR = new File(FileUtils.getTempDirectory(), "PostAggregationGapfillQueriesTest");
```
Seems like `PostAggregationGapfillQueriesTest` should be replaced with `GapfillQueriesTest`

weixiangsun replied:
PostAggregationGapfillQueriesTest is following the different query design. It will be deprecated later. I prefer to keep them as they are.

amrishlal replied:
ok

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/util/GapfillUtils.java:
```diff
@@ -31,7 +36,10 @@
  */
 public class GapfillUtils {
   private static final String POST_AGGREGATE_GAP_FILL = "postaggregategapfill";
```
Is this a left over from previous implementation? I thought `postaggregategapfill` function was no longer being used? I don't see any of the test cases using postaggregategapfill function?

weixiangsun replied:
PostAggregationGapfillQueriesTest is still using this one. I will deprecate it as part of next PR.

amrishlal replied:
ok

weixiangsun reviewed (commented):
The comments were fixed.

weixiangsun commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -369,6 +369,9 @@ private static PinotQuery compileCalciteSqlToPinotQuery(String sql) {
       DataSource dataSource = new DataSource();
       dataSource.setTableName(fromNode.toString());
       pinotQuery.setDataSource(dataSource);
+      if (fromNode instanceof SqlSelect || fromNode instanceof SqlOrderBy) {
```
I do not have deep insight into it. But the following code shows that SqlOrderBy 
 has the SqlSelect member. https://github.com/apache/pinot/blob/master/pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java#L343-L353

amrishlal replied:
OK, thanks for the reference. I would suggest changing this logic to:

```
    if (fromNode != null) {
      DataSource dataSource = new DataSource();
      if (fromNode instanceof SqlSelect || fromNode instanceof SqlOrderBy) {
        dataSource.setSubquery(compileSqlNodeToPinotQuery(fromNode));
      } else {
        dataSource.setTableName(fromNode.toString());
      }
      pinotQuery.setDataSource(dataSource);
    }

```

weixiangsun replied:
Can not leave TableName as null. I will accept your other suggestion.

amrishlal replied:
Not sure if I understand. If there is no table name  in the FROM clause, then what is the use of setting the tableName and that too to a string that is actually represents another query? Note that the definition of datasource in query.thrift file allows for tableName being null
```
struct DataSource {
  1: optional string tableName;
  2: optional PinotQuery subquery;
}
```
Also Pinot supports queries without from clause, so even in current code table name (and also subquery) can be null at the same time.

weixiangsun replied:
Here is the NullPointerException when I made your suggested change.
java.lang.NullPointerException
	at org.apache.pinot.spi.utils.builder.TableNameBuilder.tableHasTypeSuffix(TableNameBuilder.java:73)
	at org.apache.pinot.spi.utils.builder.TableNameBuilder.extractRawTableName(TableNameBuilder.java:100)
	at org.apache.pinot.core.query.reduce.BrokerReduceService.reduceOnDataTable(BrokerReduceService.java:95)
	at org.apache.pinot.queries.BaseQueriesTest.getBrokerResponse(BaseQueriesTest.java:238)
	at org.apache.pinot.queries.BaseQueriesTest.getBrokerResponseForSqlQuery(BaseQueriesTest.java:201)
	at org.apache.pinot.queries.BaseQueriesTest.getBrokerResponseForSqlQuery(BaseQueriesTest.java:173)
	at org.apache.pinot.queries.GapfillQueriesTest.datetimeconvertGapfillTestAggregateAggregateOrderBy(GapfillQueriesTest.java:3554)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:108)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:661)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:869)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1193)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:126)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:744)
	at org.testng.TestRunner.run(TestRunner.java:602)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:380)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:375)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1301)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1226)
	at org.testng.TestNG.runSuites(TestNG.java:1144)
	at org.testng.TestNG.run(TestNG.java:1115)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:109)

amrishlal replied:
My understanding is that by the time reducer code is called, you have flattened all the subqueries into a single query? In that case you would need to appropriately set a table name for reducer (or make other appropriate changes so that the query is semantically correct).

weixiangsun replied:
No, I did not flatten the subqueries into a single query.

amrishlal reviewed (approved):
It might be good to have a two or three integration test level queries as well since both broker and server are involved. Otherwise looks good to me assuming that you take care of moving the syntax checks to Broker side.

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapFillDataTableReducer.java:
```diff
@@ -0,0 +1,690 @@
+    Iterator<Record> iterator = indexedTable.iterator();
+    while (iterator.hasNext()) {
+      Object[] row = iterator.next().getValues();
+      String timeCol = row[0] instanceof Long ? ((Long) row[0]).toString() : (String) row[0];
+      long timeBucket = _dateTimeFormatter.fromFormatToMillis(timeCol);
```
Fixed

weixiangsun commented:
> It might be good to have a two or three integration test level queries as well since both broker and server are involved. Otherwise looks good to me assuming that you take care of moving the syntax checks to Broker side.

Do you have an example about how to add the integration test I can add? I do not see any query-related test against broker and server

amrishlal commented:
> Do you have an example about how to add the integration test I can add? I do not see any query-related test against broker and server

Please see `OfflineClusterIntegrationTest`

Jackie-Jiang commented:
Discussed offline. We can simplify the logic by handling the whole gapfill processing on the broker side in the `BrokerRequestHandler`:
1. When getting a gapfill query, rewrite it to the regular non-gapfill query (if the leaf subquery is gap-fill query, trim off the gap-fill part and select all the required columns)
2. Send the non-gapfill query as regular query
3. After reducing the server responses to `BrokerResponse`, apply gapfill to the `ResultTable` within the `BrokerResponse` and set the new `ResultTable`

With this logic, all the gapfill handling logic are handled at broker side in the same place without introducing unnecessary overhead to the server

weixiangsun commented:
> Discussed offline. We can simplify the logic by handling the whole gapfill processing on the broker side in the `BrokerRequestHandler`:
> 
> 1. When getting a gapfill query, rewrite it to the regular non-gapfill query (if the leaf subquery is gap-fill query, trim off the gap-fill part and select all the required columns)
> 2. Send the non-gapfill query as regular query
> 3. After reducing the server responses to `BrokerResponse`, apply gapfill to the `ResultTable` within the `BrokerResponse` and set the new `ResultTable`
> 
> With this logic, all the gapfill handling logic are handled at broker side in the same place without introducing unnecessary overhead to the server

@Jackie-Jiang I have already addressed your comments. Can you take another look?

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapfillFilterHandler.java:
```diff
@@ -0,0 +1,76 @@
+    if (expression.getType() == ExpressionContext.Type.IDENTIFIER) {
+      return new ColumnValueExtractor(_indexes.get(expression.getIdentifier()), _dataSchema);
+    } else {
+      return new ColumnValueExtractor(_indexes.get(expression.getFunction().toString()), _dataSchema);
```
This does not handle transform properly (e.g. `colA - colB` where the gapfill selects `colA` and `colB`). This is handled within the `PostAggregationValueExtractor`, and we may also extract that out to be shared. (Or add a TODO to fix later)

weixiangsun replied:
Add TODO

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/filter/RowMatcherFactory.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Factory for RowMatcher.
+ */
+public interface RowMatcherFactory {
```
This should be a concrete util class instead of an interface

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/filter/ValueExtractorFactory.java:
```diff
@@ -0,0 +1,29 @@
+ * Value extractor for the post-aggregation function or pre-aggregation gap fill.
+ */
+public interface ValueExtractorFactory {
+  ValueExtractor getValueExtractor(ExpressionContext expression);
```
Add some javadoc

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:
```diff
@@ -133,6 +137,8 @@ private QueryContext(String tableName, List<ExpressionContext> selectExpressions
     _queryOptions = queryOptions;
     _debugOptions = debugOptions;
     _brokerRequest = brokerRequest;
+    _gapfillType = null;
```
(minor) this line is redundant

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:
```diff
@@ -85,6 +86,7 @@
   // Keep the BrokerRequest to make incremental changes
   // TODO: Remove it once the whole query engine is using the QueryContext
   private final BrokerRequest _brokerRequest;
+  private final QueryContext _subQueryContext;
```
Rename it to `_subquery` to be consistent with other variables

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:
```diff
@@ -188,6 +194,10 @@ public FilterContext getHavingFilter() {
     return _orderByExpressions;
   }
 
+  public QueryContext getSubQueryContext() {
```
```suggestion
  public QueryContext getSubquery() {
```

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:
```diff
@@ -375,6 +393,57 @@ public String toString() {
     private Map<String, String> _queryOptions;
     private Map<String, String> _debugOptions;
     private BrokerRequest _brokerRequest;
+    private QueryContext _subQueryContext;
```
```suggestion
    private QueryContext _subquery;
```

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:
```diff
@@ -436,6 +505,11 @@ public Builder setBrokerRequest(BrokerRequest brokerRequest) {
       return this;
     }
 
+    public Builder setSubqueryContext(QueryContext subQueryContext) {
```
```suggestion
    public Builder setSubquery(QueryContext subquery) {
```

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:
```diff
@@ -375,6 +393,57 @@ public String toString() {
+      if (function.getType() == FunctionContext.Type.AGGREGATION) {
+        // Aggregation
+        filteredAggregations.add(Pair.of(function, null));
+      } else {
+        List<ExpressionContext> arguments = function.getArguments();
+        if (function.getFunctionName().equalsIgnoreCase("filter")) {
+          // Filtered aggregation
+          Preconditions.checkState(arguments.size() == 2, "FILTER must contain 2 arguments");
+          FunctionContext aggregation = arguments.get(0).getFunction();
+          Preconditions.checkState(aggregation != null && aggregation.getType() == FunctionContext.Type.AGGREGATION,
+              "First argument of FILTER must be an aggregation function");
+          ExpressionContext filterExpression = arguments.get(1);
+          Preconditions.checkState(filterExpression.getFunction() != null
+                  && filterExpression.getFunction().getType() == FunctionContext.Type.TRANSFORM,
+              "Second argument of FILTER must be a filter expression");
+          FilterContext filter = RequestContextUtils.getFilter(filterExpression);
+          filteredAggregations.add(Pair.of(aggregation, filter));
+        } else {
+          // Transform
+          for (ExpressionContext argument : arguments) {
+            getAggregations(argument, filteredAggregations);
+          }
+        }
+      }
+    }
+
+    /**
+     * Helper method to extract AGGREGATION FunctionContexts and FILTER FilterContexts from the given filter.
+     */
+    private static void getAggregations(FilterContext filter,
+        List<Pair<FunctionContext, FilterContext>> filteredAggregations) {
+      List<FilterContext> children = filter.getChildren();
+      if (children != null) {
+        for (FilterContext child : children) {
+          getAggregations(child, filteredAggregations);
+        }
+      } else {
+        getAggregations(filter.getPredicate().getLhs(), filteredAggregations);
+      }
+    }
```
Revert the reordering change

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/BrokerRequestToQueryContextConverter.java:
```diff
@@ -42,23 +42,42 @@
-    PinotQuery pinotQuery = brokerRequest.getPinotQuery();
-
+  private static QueryContext convertSQL(PinotQuery pinotQuery, BrokerRequest brokerRequest) {
+    QueryContext subQueryContext = null;
```
(minor)
```suggestion
    QueryContext subquery = null;
```

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/BrokerRequestToQueryContextConverter.java:
```diff
@@ -42,23 +42,42 @@
+  /**
+   * Validate the gapfill query.
+   */
+  public static void validateGapfillQuery(BrokerRequest brokerRequest) {
```
Why do we need this? This can add quite big overhead to regular queries

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorUtils.java:
```diff
@@ -88,6 +83,8 @@ private SelectionOperatorUtils() {
       ThreadLocal.withInitial(() -> new DecimalFormat(FLOAT_PATTERN, DECIMAL_FORMAT_SYMBOLS));
   private static final ThreadLocal<DecimalFormat> THREAD_LOCAL_DOUBLE_FORMAT =
       ThreadLocal.withInitial(() -> new DecimalFormat(DOUBLE_PATTERN, DECIMAL_FORMAT_SYMBOLS));
+  private SelectionOperatorUtils() {
```
Revert the unrelated changes

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorUtils.java:
```diff
@@ -391,6 +388,9 @@ public static DataTable getDataTableFromRows(Collection<Object[]> rows, DataSche
           row[i] = dataTable.getStringArray(rowId, i);
           break;
 
+        case OBJECT:
```
Does this change still apply?

weixiangsun replied:
Revert it.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/transport/QueryRouter.java:
```diff
@@ -114,6 +115,7 @@ public AsyncQueryResponse submitQuery(long requestId, String rawTableName,
     Map<ServerRoutingInstance, InstanceRequest> requestMap = new HashMap<>();
     if (offlineBrokerRequest != null) {
       assert offlineRoutingTable != null;
+      BrokerRequestToQueryContextConverter.validateGapfillQuery(offlineBrokerRequest);
```
Let's not validate gapfill here as it will add overhead to all queries (converting broker request to query context)

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/util/GapfillUtils.java:
```diff
@@ -71,12 +86,15 @@ public static boolean isFill(ExpressionContext expressionContext) {
     }
 
-    return FILL.equals(canonicalizeFunctionName(expressionContext.getFunction().getFunctionName()));
+    return FILL.equalsIgnoreCase(canonicalizeFunctionName(expressionContext.getFunction().getFunctionName()));
```
After #8341, all function name in `FunctionContext` is already canonical, no need to canonicalize again. Same for other places
```suggestion
    return expressionContext.getFunction().getFunctionName().equals(FILL);
```

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/util/GapfillUtils.java:
```diff
@@ -119,4 +137,265 @@ static public Serializable getDefaultValue(DataSchema.ColumnDataType dataType) {
+    if (brokerRequest.getPinotQuery().getDataSource() == null) {
+      return brokerRequest;
+    }
+    QueryContext queryContext = BrokerRequestToQueryContextConverter.convert(brokerRequest);
```
Avoid this conversion. We may loop over the select list and see if there is `gapfill`, and directly rewrite the query when `gapfill` is found

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/util/GapfillUtils.java:
```diff
@@ -119,4 +137,265 @@ static public Serializable getDefaultValue(DataSchema.ColumnDataType dataType) {
+      if (select.getType() != ExpressionType.FUNCTION) {
+        continue;
+      }
+      if (GAP_FILL.equalsIgnoreCase(select.getFunctionCall().getOperator())) {
```
(minor) function name is canonical, so you may use `equals` here

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/util/GapfillUtils.java:
```diff
@@ -119,4 +137,265 @@ static public Serializable getDefaultValue(DataSchema.ColumnDataType dataType) {
+    return brokerRequest;
+  }
+
+  public enum GapfillType {
```
Going over all the classes, I feel `GapfillType` might not be required at all. All the handling can be based on levels of subqueries.
If you feel `GapfillType` can make code cleaner, we may add a function `GapfillType getGapfillType(QueryContext queryContext)` and maintain it within the `GapfillProcessor`. No need to embed it into the `QueryContext` (in `QueryContext`, we try to maintain only common properties, not feature specific ones).

weixiangsun replied:
Fixed.

Jackie-Jiang reviewed (commented):
Mostly good. Good job extracting the common matcher and extractor code, and implement the row based `BlockValSet`. All the gapfill logic now resides in a few classes, which is much cleaner.
Let's try to make the overhead to non-gapfill queries minimal

weixiangsun commented on pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:
```diff
@@ -217,6 +218,10 @@ private BrokerResponseNative handleSQLRequest(long requestId, String query, Json
     }
+
+    BrokerRequest originalBrokerRequest = brokerRequest;
+    brokerRequest = GapfillUtils.stripGapfill(originalBrokerRequest);
```
Fixed

weixiangsun commented on pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:
```diff
@@ -2183,9 +2192,9 @@ private void attachTimeBoundary(String rawTableName, BrokerRequest brokerRequest
-      @Nullable BrokerRequest offlineBrokerRequest, @Nullable Map<ServerInstance, List<String>> offlineRoutingTable,
-      @Nullable BrokerRequest realtimeBrokerRequest, @Nullable Map<ServerInstance, List<String>> realtimeRoutingTable,
-      long timeoutMs, ServerStats serverStats, RequestStatistics requestStatistics)
+      BrokerRequest brokerRequest, @Nullable BrokerRequest offlineBrokerRequest, @Nullable Map<ServerInstance,
```
Fixed

weixiangsun commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -67,9 +67,7 @@
-  private CalciteSqlParser() {
-  }
-
+  public static final List<QueryRewriter> QUERY_REWRITERS = new ArrayList<>(QueryRewriterFactory.getQueryRewriters());
```
Fixed

weixiangsun commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -100,6 +95,9 @@ private CalciteSqlParser() {
   private static final Pattern OPTIONS_REGEX_PATTEN =
       Pattern.compile("option\\s*\\(([^\\)]+)\\)", Pattern.CASE_INSENSITIVE);
 
+  private CalciteSqlParser() {
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/BrokerReduceService.java:
```diff
@@ -108,7 +108,12 @@ public BrokerResponseNative reduceOnDataTable(BrokerRequest brokerRequest,
         new DataTableReducerContext(_reduceExecutorService, _maxReduceThreadsPerQuery, reduceTimeOutMs,
             _groupByTrimThreshold), brokerMetrics);
-    updateAlias(queryContext, brokerResponseNative);
+    QueryContext originalQueryContext = BrokerRequestToQueryContextConverter.convert(originalBrokerRequest);
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/ColumnDataToBlockValSetConverter.java:
```diff
@@ -0,0 +1,181 @@
+ * block of values {@link BlockValSet} which used as input to the aggregation
+ * function.
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/ColumnDataToBlockValSetConverter.java:
```diff
@@ -0,0 +1,181 @@
+ * function.
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class ColumnDataToBlockValSetConverter implements BlockValSet {
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapFillProcessor.java:
```diff
@@ -0,0 +1,455 @@
+ * Helper class to reduce and set gap fill results into the BrokerResponseNative
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class GapFillProcessor {
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/ColumnDataToBlockValSetConverter.java:
```diff
@@ -0,0 +1,181 @@
+  @Nullable
+  @Override
+  public Dictionary getDictionary() {
+    throw new UnsupportedOperationException("Not supported");
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/ColumnDataToBlockValSetConverter.java:
```diff
@@ -0,0 +1,181 @@
+
+  @Override
+  public int[] getDictionaryIdsSV() {
+    throw new UnsupportedOperationException("Not supported");
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/ColumnDataToBlockValSetConverter.java:
```diff
@@ -0,0 +1,181 @@
+
+  @Override
+  public int[] getIntValuesSV() {
+    if (_dataType == FieldSpec.DataType.INT) {
+      int[] result = new int[_rows.size()];
+      for (int i = 0; i < result.length; i++) {
+        result[i] = (Integer) _rows.get(i)[_columnIndex];
+      }
+      return result;
+    }
+    throw new UnsupportedOperationException("Not supported");
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/ColumnDataToBlockValSetConverter.java:
```diff
@@ -0,0 +1,181 @@
+
+
+/**
+ * As for Gapfilling Function, all raw data will be retrieved from the pinot
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/ColumnDataToBlockValSetConverter.java:
```diff
@@ -0,0 +1,181 @@
+
+  @Override
+  public String[] getStringValuesSV() {
+    if (_dataType == FieldSpec.DataType.STRING) {
+      String[] result = new String[_rows.size()];
+      for (int i = 0; i < result.length; i++) {
+        result[i] = (String) _rows.get(i)[_columnIndex];
+      }
+      return result;
+    }
+    throw new UnsupportedOperationException("Not supported");
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/filter/ValueExtractorFactory.java:
```diff
@@ -0,0 +1,29 @@
+
+
+/**
+ * Value extractor for the post-aggregation function or pre-aggregation gap fill.
```
Fixed

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapfillFilterHandler.java:
```diff
@@ -0,0 +1,76 @@
+    _dataSchema = dataSchema;
+    _indexes = new HashMap<>();
+    for (int i = 0; i < _dataSchema.size(); i++) {
+      _indexes.put(_dataSchema.getColumnName(i), i);
```
Add TODO

weixiangsun commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapFillProcessor.java:
```diff
@@ -0,0 +1,455 @@
+      }
+    }
+
+    if (_gapfillType == GapfillUtils.GapfillType.GAP_FILL_AGGREGATE || _gapfillType == GapfillUtils.GapfillType.GAP_FILL
+        || _gapfillType == GapfillUtils.GapfillType.GAP_FILL_SELECT) {
+      List<Object[]> gapfilledRows = gapFillAndAggregate(timeBucketedRawRows, resultTableSchema, dataSchema);
+      if (_gapfillType == GapfillUtils.GapfillType.GAP_FILL_SELECT) {
+        resultRows = new ArrayList<>(gapfilledRows.size());
+        resultRows.addAll(gapfilledRows);
+      } else {
+        resultRows = gapfilledRows;
+      }
+    } else {
+      resultRows = gapFillAndAggregate(timeBucketedRawRows, resultTableSchema, dataSchema);
+    }
```
Good catch, Fixed.

weixiangsun commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java:
```diff
@@ -162,8 +163,7 @@ public String getQueryResponse(String query, String traceEnabled, String queryOp
         case CommonConstants.Broker.Request.SQL:
-          inputTableName =
-              SQL_QUERY_COMPILER.compileToBrokerRequest(query).getPinotQuery().getDataSource().getTableName();
+          inputTableName = GapfillUtils.getTableName(SQL_QUERY_COMPILER.compileToBrokerRequest(query).getPinotQuery());
```
Fixed.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/BrokerReduceService.java:
```diff
@@ -103,11 +104,23 @@ public BrokerResponseNative reduceOnDataTable(BrokerRequest brokerRequest,
         new DataTableReducerContext(_reduceExecutorService, _maxReduceThreadsPerQuery, reduceTimeOutMs,
             _groupByTrimThreshold), brokerMetrics);
+    QueryContext queryContext;
+    if (brokerRequest == serverBrokerRequest) {
+      queryContext = serverQueryContext;
+    } else {
+      queryContext = BrokerRequestToQueryContextConverter.convert(brokerRequest);
+    }
+
+    GapfillUtils.GapfillType gapfillType = GapfillUtils.getGapfillType(queryContext);
+    if (gapfillType != null) {
+      GapfillProcessor gapfillProcessor = new GapfillProcessor(queryContext, gapfillType);
+      gapfillProcessor.process(brokerResponseNative);
+    }
```
No need to check gapfill type when server request is the same as broker request
```suggestion
    if (brokerRequest == serverBrokerRequest) {
      queryContext = serverQueryContext;
    } else {
      queryContext = BrokerRequestToQueryContextConverter.convert(brokerRequest);
      GapfillUtils.GapfillType gapfillType = GapfillUtils.getGapfillType(queryContext);
      if (gapfillType != null) {
        GapfillProcessor gapfillProcessor = new GapfillProcessor(queryContext, gapfillType);
        gapfillProcessor.process(brokerResponseNative);
      }
    }
```

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/test/java/org/apache/pinot/queries/BaseQueriesTest.java:
```diff
@@ -197,21 +198,31 @@ protected BrokerResponseNative getBrokerResponseForSqlQuery(String sqlQuery, Pla
     }
     queryOptions.put(Request.QueryOptionKey.GROUP_BY_MODE, Request.SQL);
     queryOptions.put(Request.QueryOptionKey.RESPONSE_FORMAT, Request.SQL);
+    BrokerRequest strippedBrokerRequest = GapfillUtils.stripGapfill(brokerRequest);
```
(MAJOR) Query options should be preserved when stripping the gapfill

Jackie-Jiang replied:
(minor) rename it to `serverBrokerRequest`

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:
```diff
@@ -217,7 +218,10 @@ private BrokerResponseNative handleSQLRequest(long requestId, String query, Json
     }
-    PinotQuery pinotQuery = brokerRequest.getPinotQuery();
+
+    BrokerRequest serverBrokerRequest = GapfillUtils.stripGapfill(brokerRequest);
```
(MAJOR) Let's set the query options first (some options are already set during the query compilation), then strip the gapfill. You may just set the stripped query options to be the original query options without making a copy

weixiangsun replied:
It is done inside GapfillUtils.stripGapfill method.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/filter/RowMatcherFactory.java:
```diff
@@ -0,0 +1,47 @@
+        return new AndRowMatcher(filter.getChildren(), valueExtractorFactory);
+      case OR:
+        return new OrRowMatcher(filter.getChildren(), valueExtractorFactory);
+      case PREDICATE:
```
We need to handle `NOT` here. This is already fixed in #8366. If that one is merged first, we should integrate the fix here.

weixiangsun replied:
I prefer to integrate the fix here until your fix is merged in.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorUtils.java:
```diff
@@ -79,7 +79,6 @@ private SelectionOperatorUtils() {
   private static final String FLOAT_PATTERN = "#########0.0####";
   private static final String DOUBLE_PATTERN = "###################0.0#########";
   private static final DecimalFormatSymbols DECIMAL_FORMAT_SYMBOLS = DecimalFormatSymbols.getInstance(Locale.US);
-
```
(minor) revert this file

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/test/java/org/apache/pinot/queries/BaseQueriesTest.java:
```diff
@@ -197,21 +198,31 @@ protected BrokerResponseNative getBrokerResponseForSqlQuery(String sqlQuery, Pla
+    queryOptions.put(Request.QueryOptionKey.RESPONSE_FORMAT, Request.SQL);
     QueryContext queryContext = BrokerRequestToQueryContextConverter.convert(brokerRequest);
-    return getBrokerResponse(queryContext, planMaker);
+    QueryContext strippedQueryContext = BrokerRequestToQueryContextConverter.convert(strippedBrokerRequest);
```
Let's compare the reference before converting the `strippedBrokerRequest`, and rename it to `serverQueryContext`. Same for `getBrokerResponseForOptimizedSqlQuery`

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/test/java/org/apache/pinot/queries/BaseQueriesTest.java:
```diff
@@ -197,21 +198,31 @@ protected BrokerResponseNative getBrokerResponseForSqlQuery(String sqlQuery, Pla
    */
-  private BrokerResponseNative getBrokerResponse(QueryContext queryContext, PlanMaker planMaker) {
+  private BrokerResponseNative getBrokerResponse(
+      QueryContext queryContext, QueryContext strippedQueryContext, PlanMaker planMaker) {
```
```suggestion
      QueryContext queryContext, QueryContext serverQueryContext, PlanMaker planMaker) {
```

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapfillProcessor.java:
```diff
@@ -0,0 +1,471 @@
+          }
+          result.add(resultRow);
+        }
+        bucketedResult = new ArrayList<>();
```
Unnecessary allocation
```suggestion
        bucketedResult = bucketedResult.clear();
```

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapfillProcessor.java:
```diff
@@ -0,0 +1,471 @@
+            return result;
+          }
+        }
+        bucketedResult = new ArrayList<>();
```
```suggestion
        bucketedResult.clear();
```

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapfillProcessor.java:
```diff
@@ -0,0 +1,471 @@
+          result.add(resultRow);
+        }
+        bucketedResult = new ArrayList<>();
+      } else if (index % _aggregationSize == _aggregationSize - 1 && bucketedResult.size() > 0) {
```
Can `bucketedResult` ever be empty? If it is empty, do we need to update the `start`?

weixiangsun replied:
Yes, we need update the start if it is empty.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapfillProcessor.java:
```diff
@@ -0,0 +1,471 @@
+
+    _dateTimeFormatter = new DateTimeFormatSpec(args.get(1).getLiteral());
+    _gapfillDateTimeGranularity = new DateTimeGranularitySpec(args.get(4).getLiteral());
+    _postGapfillDateTimeGranularity = new DateTimeGranularitySpec(args.get(5).getLiteral());
```
Trying to understand when we need to use different granularity for gapfill and post-gapfill. Does this align with the gapfill definition?

weixiangsun replied:
Here is background information.

Currently we only have one time bucket size. Usually one entity only has one representative state for one time bucket. Take the parking lot as an example, one parking lot can be occupied or not.

If we want to know how long the park lots are occupied per day inside the parking building. The time bucket size is one day. The parking lot state will be decided by one event within the time bucket. 

By introducing the different granularity for gapfill and post-gapfill, it will make it possible to calculate the above metrics more precisely. For parking lot, we can introduce 5 minutes as gapfill bucket size, then aggregate all occupied states within post-gapfill time bucket (1 day).

Please let me know if you have any question about it. @Jackie-Jiang

Jackie-Jiang replied:
My concern is that this seems not align with the gapfill definition. Currently the bucketed rows are passed to the parent aggregation function to be aggregated, but conceptually there should be no dependency between parent query and subquery, and parent query should just run on top of subquery results.
If this functionality is a must have, we may keep the workaround, but let's make this argument optional so that it is easy to use. I find it quite hard to understand the semantic of this post-gapfill granularity

Jackie-Jiang commented on pinot-core/src/test/java/org/apache/pinot/queries/GapfillQueriesTest.java:
```diff
@@ -0,0 +1,3701 @@
+        + " GROUP BY time_col "
+        + " LIMIT 2000 ";
+
+    System.out.println(gapfillQuery1);
```
Remove the console print

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapfillProcessor.java:
```diff
@@ -0,0 +1,473 @@
+          new RowBasedBlockValSet(dataSchema.getColumnDataType(i), bucketedRows, i));
+    }
+
+    for (int i = 0; i < _queryContext.getSelectExpressions().size(); i++) {
```
You should be able to use the `_queryContext.getAggregationFunctions()`

weixiangsun replied:
I need the selection index of aggregationFunction.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapfillProcessor.java:
```diff
@@ -0,0 +1,477 @@
+      int index = findGapfillBucketIndex(time);
+      gapfill(time, bucketedResult, timeBucketedRawRows[index], dataSchema, postGapfillFilterHandler);
+      if (_queryContext.getAggregationFunctions() == null) {
+        for (Object [] row : bucketedResult) {
```
(code format) reformat this file

weixiangsun replied:
Fixed

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapfillProcessor.java:
```diff
@@ -0,0 +1,477 @@
+    for (long time = _startMs; time < _endMs; time += _gapfillTimeBucketSize) {
+      int index = findGapfillBucketIndex(time);
+      gapfill(time, bucketedResult, timeBucketedRawRows[index], dataSchema, postGapfillFilterHandler);
+      if (_queryContext.getAggregationFunctions() == null) {
```
We may simplify the same granularity case
```suggestion
      if (_queryContext.getAggregationFunctions() == null || _aggregationSize == 1) {
```

weixiangsun replied:
When _queryContext.getAggregationFunctions() == null, the bucketed result will be added to final result. But when _aggregationSize == 1, it need go through the aggregation.

So I cannot take this suggestion.

Jackie-Jiang reviewed (approved):
LGTM
